### PR TITLE
feat(telemetry): update telemetry to report metrics

### PR DESCRIPTION
**Pull request by avara1986**
# Description
Split telemetry worker: One Worker to report events (heartbeat, startup...) Other to report metrics.
Class methods to report distributions, gauge, count and rate Metrics type. 
ASM Metrics such as `waf.requests`, `waf.duration`, etc.

# Motivation
ASM needs to report information about 

# Testing strategy
See Slack

# PRs related:
System tests: https://github.com/DataDog/system-tests/pull/823

This PR comes from https://github.com/DataDog/dd-trace-py/pull/4943 cc @mabdinur 

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.

### CODE REVIEW

File: ddtrace/internal/telemetry/metrics.py
Comment (christophe-papazian) [inline ddtrace/internal/telemetry/metrics.py:95]: maybe remove the comment lines that are not used anymore ?

File: ddtrace/internal/telemetry/metrics_namespaces.py
Comment (christophe-papazian) [inline ddtrace/internal/telemetry/metrics_namespaces.py:35]: why checking only the class ?

File: ddtrace/internal/telemetry/metrics_namespaces.py
Comment (christophe-papazian) [inline ddtrace/internal/telemetry/metrics_namespaces.py:58]: Maybe use set instead of dict for namespace, as the key here is meaningless, and sets will use the id from hash anyway.

Comment (avara1986) [reply ddtrace/internal/telemetry/metrics_namespaces.py:58]: Each metric is sent in a single payload group by namespace (appsec, tracer...) we use it to aggragate the metrics for each namespace :+1:

File: ddtrace/internal/telemetry/metrics_namespaces.py
Comment (christophe-papazian) [reply ddtrace/internal/telemetry/metrics_namespaces.py:35]: ok, but the way it is done here is based on id, not name. So two metrics with the same name but different tags will still be possible. is this ok?

Comment (avara1986) [reply ddtrace/internal/telemetry/metrics_namespaces.py:35]: yes, two metrics with the same name but different tags it will be reported as two different metrics (check it https://www.datadoghq.com/blog/the-power-of-tagged-metrics/#whats-a-metric-tag)

But we CAN NOT have two metrics with the same name and different type

Comment (christophe-papazian) [reply ddtrace/internal/telemetry/metrics_namespaces.py:36]: ok, this looks perfect!

**Review by christophe-papazian — DISMISSED** [review]

File: ddtrace/internal/telemetry/metrics_namespaces.py
Comment (mabdinur) [inline ddtrace/internal/telemetry/metrics_namespaces.py:51]: Is this error caught by the telemetry writer? Could this error crash an application?

Comment (avara1986) [reply ddtrace/internal/telemetry/metrics_namespaces.py:51]: This error raises when you call `telemetry_writer.add_[type]_metric()`. In this case, a crash error it's ok to catch the problem while we're developing. If not, we've the risk to add a metric with errors/not working, merge and add it into a release

File: ddtrace/internal/constants.py
Comment (emmettbutler) [inline ddtrace/internal/constants.py:21]: A slightly more descriptive naming convention on these constants would help readers understand their intended purpose. `TELEMETRY_TAG_TRACER`, maybe?

File: ddtrace/appsec/processor.py
Comment (emmettbutler) [inline ddtrace/appsec/processor.py:302]: What's the plan with these commented lines?

File: tests/telemetry/test_telemetry_metrics.py
Comment (emmettbutler) [inline tests/telemetry/test_telemetry_metrics.py:33]: This docstring didn't help me understand the function. You might consider removing it here and in the few other places it's duplicated.

File: tests/telemetry/test_telemetry_metrics.py
Comment (emmettbutler) [inline tests/telemetry/test_telemetry_metrics.py:56]: I don't understand how this assertion maps onto the idea of "flush and series is restarted". Is the key that the `points` array only has one element in it, indicating that another datapoint was not added to the series?

File: tests/telemetry/test_telemetry_metrics.py
Comment (emmettbutler) [inline tests/telemetry/test_telemetry_metrics.py:60]:
```suggestion
def test_send_metric_datapoint_equal_type_and_tags_yields_single_series(test_agent_session_telemetry_metrics, mock_time):
```

File: tests/telemetry/test_telemetry_metrics.py
Comment (emmettbutler) [inline tests/telemetry/test_telemetry_metrics.py:85]:
```suggestion
def test_send_metric_datapoint_equal_type_different_tags_yields_multiple_series(
```

File: tests/telemetry/test_telemetry_metrics.py
Comment (emmettbutler) [inline tests/telemetry/test_telemetry_metrics.py:142]: I think this assertion needs to be within the `with` block to see the `e` variable. It's at least slightly clearer to the reader that way.

File: tests/telemetry/test_telemetry_metrics.py
Comment (emmettbutler) [inline tests/telemetry/test_telemetry_metrics.py:130]:
```suggestion
def test_send_metric_datapoint_equal_tags_different_type_throws_error(
```

File: ddtrace/internal/telemetry/metrics_namespaces.py
Comment (emmettbutler) [inline ddtrace/internal/telemetry/metrics_namespaces.py:25]: `reset` might be a more appropriate name for this functionality, if I understand correctly what it's doing.

File: ddtrace/internal/telemetry/metrics.py
Comment (emmettbutler) [inline ddtrace/internal/telemetry/metrics.py:21]: What does this mean?

File: ddtrace/internal/telemetry/metrics.py
Comment (emmettbutler) [inline ddtrace/internal/telemetry/metrics.py:30]: This comment doesn't add any information one can't get from the variable name. I'd probably remove it.

File: ddtrace/internal/telemetry/metrics.py
Comment (emmettbutler) [inline ddtrace/internal/telemetry/metrics.py:31]: Maybe you could name this argument `is_common_to_all_tracers`.

File: ddtrace/internal/telemetry/metrics.py
Comment (emmettbutler) [inline ddtrace/internal/telemetry/metrics.py:37]:
```suggestion
        self.interval = self._roll_up_interval = interval
```

File: ddtrace/internal/telemetry/metrics.py
Comment (emmettbutler) [inline ddtrace/internal/telemetry/metrics.py:59]: It might not be particularly intuitive to a caller that this function replaces `self._tags` rather than `extend`ing it. You might either clearly document that fact or change the implementation to use `extend`.

File: ddtrace/internal/telemetry/metrics.py
Comment (emmettbutler) [inline ddtrace/internal/telemetry/metrics.py:112]: Looks like this replaces the entire `_points` array with a single-element array. Why is that?

**Review by emmettbutler — CHANGES_REQUESTED** [review]
Looking good, @avara1986! I took a close look at this today because the Python APM group was tagged for a review. I tried to be as helpful as possible in my comments. Please ask any questions you have and tell me if you'd like me to make any adjustments to my feedback approach. Thank you for the pull request!

File: ddtrace/internal/constants.py
Comment (avara1986) [reply ddtrace/internal/constants.py:21]: You're right, I think the correct name would be `TELEMETRY_NAMESPACE_TAG_TRACER`

File: ddtrace/appsec/processor.py
Comment (avara1986) [reply ddtrace/appsec/processor.py:302]: I added a TODO below:
https://github.com/DataDog/dd-trace-py/pull/5046/files/49c7b6f056188c30450615d1349b0165e161309a#diff-0b32c6cdae8bdd01786317bcabebf73dbb57e535adab74c94858fda05da411efR610

I'm going to add the same TODO here

File: ddtrace/internal/telemetry/metrics_namespaces.py
Comment (avara1986) [reply ddtrace/internal/telemetry/metrics_namespaces.py:25]: I set _flush name because TelemetryWriter use the same terms to retrieve and reset the data

File: ddtrace/internal/telemetry/writer.py
Comment (avara1986) [reply ddtrace/internal/telemetry/writer.py:78]: Ups, I think it was a wrong automatic merge :S

File: ddtrace/internal/telemetry/metrics.py
Comment (avara1986) [reply ddtrace/internal/telemetry/metrics.py:99]: DD developers. If we want to add new metrics I think it will useful to have got the description of the type of metrics close :+1:

File: tests/telemetry/test_telemetry_metrics.py
Comment (avara1986) [reply tests/telemetry/test_telemetry_metrics.py:55]: When we add a metric, this data is stored in the telemetry_writer instance. When we run telemetry_writer.periodic() the information is sent to the agent and flush the queue.

We want to test isolated elements in the queue and ensure the queue is empty

File: ddtrace/internal/telemetry/metrics.py
Comment (avara1986) [reply ddtrace/internal/telemetry/metrics.py:111]: cause GaugeMetric we want to store the last measure we take

File: ddtrace/internal/constants.py
Comment (gnufede) [inline ddtrace/internal/constants.py:19]: You're removing a comment here

File: ddtrace/internal/telemetry/metrics.py
Comment (avara1986) [reply ddtrace/internal/telemetry/metrics.py:22]: e.g: https://docs.datadoghq.com/metrics/#what-are-metrics

**Review by juanjux — APPROVED** [review]

**Review by gnufede — APPROVED** [review]

File: ddtrace/internal/constants.py
Comment (mabdinur) [inline ddtrace/internal/constants.py:19]:
```suggestion
```

File: ddtrace/internal/remoteconfig/client.py
Comment (mabdinur) [inline ddtrace/internal/remoteconfig/client.py:28]: Is this change required?

File: ddtrace/internal/telemetry/metrics.py
Comment (mabdinur) [inline ddtrace/internal/telemetry/metrics.py:98]: _points is an attribute in the parent class (Metric). We shouldn't override it here.

```suggestion
```

File: ddtrace/internal/telemetry/metrics.py
Comment (mabdinur) [inline ddtrace/internal/telemetry/metrics.py:157]: _points is an attribute in the parent class (Metric). We shouldn't override it here.

```suggestion
```

File: tests/telemetry/conftest.py
Comment (mabdinur) [inline tests/telemetry/conftest.py:115]: We should update telemetry_writer._reset() to flush metrics queues

```suggestion
    test_agent_session.telemetry_writer._reset()
```

File: tests/telemetry/test_writer.py
Comment (mabdinur) [inline tests/telemetry/test_writer.py:32]: Can we add test coverage for the new headers telemetry headers

File: ddtrace/internal/telemetry/writer.py
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:91]: api version is also defined in the request body, it might make sense to keep both values consistent. 

```suggestion
        headers["DD-Telemetry-API-Version""] = request["api_version"]
        headers["DD-Agent-Hostname"] = get_host_info()["hostname"]
```

File: ddtrace/internal/telemetry/writer.py
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:243]: api version on line 255 should be v2

**Review by mabdinur — COMMENTED** [review]
Called out some nits, but overall looks good to me. Seeing how this is a high impact change can we run this branch on a sample app and see if we collect accurate metrics or release this feature behind a feature flag? 

The instrumentation telemetry datadog org will aggregate metrics across tracer versions. If the first release of this feature submits inaccurate metrics we can expect to receive these values for the foreseeable future.

File: tests/telemetry/test_writer.py
Comment (avara1986) [reply tests/telemetry/test_writer.py:32]: :+1:

File: ddtrace/internal/telemetry/writer.py
Comment (avara1986) [reply ddtrace/internal/telemetry/writer.py:96]: :ok_hand:

File: ddtrace/internal/telemetry/metrics.py
Comment (avara1986) [reply ddtrace/internal/telemetry/metrics.py:157]: :+1:

File: ddtrace/internal/remoteconfig/client.py
Comment (avara1986) [reply ddtrace/internal/remoteconfig/client.py]: No, it isn't, rollback

File: ddtrace/internal/telemetry/writer.py
Comment (avara1986) [reply ddtrace/internal/telemetry/writer.py]: :ok_hand:

File: tests/telemetry/conftest.py
Comment (avara1986) [reply tests/telemetry/conftest.py]: Updated with the last refactor, thanks for the suggestion!

File: ddtrace/internal/remoteconfig/client.py
Comment (avara1986) [reply ddtrace/internal/remoteconfig/client.py]: :+1:

File: ddtrace/internal/telemetry/writer.py
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:144]: Something to call out. I spoke to @mehulsonowal and it appears api v2 is not fully backwards compatible with v1. We should avoid bumping the telemetry writer to v2 in this PR (if possible) or scope api v2 to just metrics events.

Comment (avara1986) [reply ddtrace/internal/telemetry/writer.py:144]: yes, I noticed if I update all telemetry requests to v2 system tests start to fail. I rollback to v1 and I tested manually and everything works fine. On the other hand, I split TelemetryWriter into TelemetryWriter (inverval of 60s and we use it to report heartbeat, startapp, close...) and TelemetryMetricsWriter (interval of 10s and to report generate-metrics and distributions). 

Thank you so much for your tip!

File: .github/CODEOWNERS
Comment (gnufede) [inline .github/CODEOWNERS:30]: I think this should be covered by this line below:
```
tests/contrib/*/test*appsec*.py @DataDog/asm-python
```

Comment (avara1986) [reply .github/CODEOWNERS]: :+1:

**Review by juanjux — DISMISSED** [review]

File: ddtrace/appsec/iast/cmake-build-debug/CMakeFiles/clion-log.txt
Comment (mabdinur) [inline ddtrace/appsec/iast/cmake-build-debug/CMakeFiles/clion-log.txt:1]: Should we remove this file from this PR and add it to `.gitignore`?

File: ddtrace/appsec/processor.py
Comment (mabdinur) [inline ddtrace/appsec/processor.py:195]: Ideally this PR should introduce the functionality to queue and send telemetry metrics. Then in a follow up PR we can use the telemetry writer to send metrics in different products. Can we decouple adding the telemetry metrics writer change from sending telemetry metrics in ASM products? 

This way if there is an issue with the telemetry writer (performance or a bug) we can just revert/fix the PRs which send telemetry and not all the logic in this PR. This is just a suggestion, if we need this functionality merged ASAP then I am fine with leaving this PR as is

Comment (mabdinur) [inline ddtrace/appsec/processor.py:195]: We should add test coverage for test metrics. Right now there is no regression tests for appsec processor metrics, the values can change/break at any point and we won't be notified.

Comment (avara1986) [reply ddtrace/appsec/processor.py:195]: Thank you so much @mabdinur , Nice idea to add performance tests :+1: 
About integration and regression tests is it enough with these tests?
- https://github.com/DataDog/dd-trace-py/pull/5046/files#diff-e56afc084b0d6a7e0a5d1c16ead0bc8d9916c99127c9c1d2224130b9e940e11eR1
- https://github.com/DataDog/dd-trace-py/pull/5046/files#diff-c210bd10d4e51b5ab136feb9a72477acc700d4183c3c59da7b942fc0de6d5d53R1
- https://github.com/DataDog/system-tests/pull/890

DM me if you want to share your thoughts and discuss the best approach :upside_down_face:

Comment (avara1986) [reply ddtrace/appsec/processor.py:195]: Thanks! your suggestion is a good approach! If I start now to split the PR in smaller PRs I'll became mad :sweat_smile: What do you think about this idea?
- I'm continue in Telemetry Metrics in this branch
- When we're ready for the final review and merge I can split this PR into "Telemetry Metrics Core" and "ASM Telemetry Metrics"

is it OK for you? I'd love to know if this approach is quite so smooth for all of us :+1:

Comment (mabdinur) [reply ddtrace/appsec/processor.py:195]: Yup that sounds great to me, it would be nice to get @Yun-Kim feedback on the performance benchmarks. Thanks for sticking with this. This PR looks good to me. If you can get another thumbs up from another review then we can split up and merging this change.

**Comment by mabdinur** [comment]
LGTM, thanks for sticking with this 😄

File: ddtrace/appsec/processor.py
Comment (Yun-Kim) [reply ddtrace/appsec/processor.py:195]: I'm curious why this change results in a regression in the `span-add-metrics` microbenchmark scenario, which specifically only call `span.set_metrics()` and shouldn't be impacted by telemetry, unless I'm missing something. @avara1986 do you have any clues as to why we're seeing this?

Comment (avara1986) [reply ddtrace/appsec/processor.py:195]: ey Yun! no, I haven't. It's strange because the microbenchmark didn't report anything in the last commits, even after adding specific benchmark tests for Telemetry Metrics. `span-add-metrics` reports a regression after a merge of 1.x in this branch. But I merged again 1.x into this branch and the regression alert disappear. Flaky? I figure out that to be sure :+1:

**Review by emmettbutler — APPROVED** [review]

**Pull request closed without merging**
